core/services/eth/errors.go

```diff
@@ -227,3 +234,40 @@ func isParityFatal(s string) bool {
 			parInvalidSignature.MatchString(s) ||
 			parInvalidRlp.MatchString(s))
 }
+
+// ExtractRevertReasonFromRPCError attempts to extract the revert reason from the response of
+// an RPC eth_call that reverted by parsing the message from the "data" field
+// ex:
+// kovan (parity)
+// { "error": { "code" : -32015, "data": "Reverted 0xABC123...", "message": "VM execution error." } } // revert reason always omitted
+// rinkeby / ropsten (geth)
+// { "error":  { "code": 3, "data": "0x0xABC123...", "message": "execution reverted: hello world" } } // revert reason included in message
+func ExtractRevertReasonFromRPCError(err error) (string, error) {
+	if err == nil {
+		return "", errors.New("no error present")
+	}
+	cause := errors.Cause(err)
+	field := reflect.ValueOf(cause).Elem().FieldByName("Data")
```

Comment:
I think we could just json unmarshal here instead of reflection and avoid the panic risk? If we have to use reflection then we should add a defer&recover

Reply:
:+1: